### PR TITLE
Server: revisit server exceptions

### PR DESCRIPTION
**Pull request by dmarteau**
This PR brings some changes on how to define exceptions in qgis server: 

- Add server exception base class for a more homogoneous exception handling. 
- Enable per service exception definition: each service may define the response format (handle OGC exception versioning that depends on the invoked service)
- Handle QgsException gracefully (error 500)

### CODE REVIEW

File: src/server/qgsserver.cpp [hunk 1]
```diff
@@ -387,109 +387,121 @@ void QgsServer::handleRequest( QgsServerRequest& request, QgsServerResponse& res
...
 
-  //Config file path
-  QString configFilePath = configPath( *sConfigFilePath, parameterMap );
+      //Config file path
+      QString configFilePath = configPath( *sConfigFilePath, parameterMap );
 
 #ifdef HAVE_SERVER_PYTHON_PLUGINS
-  // XXX Why this is enabled only fol plugins ?
-  sServerInterface->setConfigFilePath( configFilePath );
+      // XXX Why this is enabled only for plugins ?
```
Comment (elpaso) [inline src/server/qgsserver.cpp]: because server interface does not exist if plugins are disabled

Comment (dmarteau) [inline src/server/qgsserver.cpp]: So this directive has no purpose now:   server interface always exists now  (passed to service at init time).

File: src/server/qgsmapserverutils.cpp [hunk 2]
```diff
@@ -1,84 +0,0 @@
...
- ***************************************************************************/
-#include "qgsmapserverutils.h"
-#include "qgslogger.h"
-#include <stdlib.h>
-#include <time.h>
-#include <QDir>
-#include <QFileInfo>
-#include <QTextStream>
-
-QString QgsMapServerUtils::createTempFilePath()
```
Comment (nyalldawson) [inline src/server/qgsmapserverutils.cpp]: I'm very glad to see these methods go - Coverity has been flagging them as unsafe for as long as we've been running scan over QGIS!

File: src/server/qgsmapserviceexception.h [hunk 3]
```diff
@@ -34,15 +36,12 @@
...
+class SERVER_EXPORT QgsMapServiceException : public QgsOgcServiceException
 {
   public:
-    QgsMapServiceException( const QString& code, const QString& message );
-    QString code() const {return mCode;}
-    QString message() const {return mMessage;}
-  private:
-    QString mCode;
-    QString mMessage;
+    QgsMapServiceException( const QString& code, const QString& message )
```
Comment (nyalldawson) [inline src/server/qgsmapserviceexception.h]: Why not just remove this class totally and document the api break in doc/api_break.dox?

File: python/server/qgsserverexception.sip [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+{
+%TypeHeaderCode
+#include <qgsserverexception.h>
+%End
+  public:
+    QgsServerException( const QString& message, int responseCode = 500 );
+
+    int responseCode() const;
+
+    virtual QByteArray formatResponse( QString& responseFormat ) const;
```
Comment (nyalldawson) [inline python/server/qgsserverexception.sip]: responseFormat should be marked with an /Out/ annotation:

    virtual QByteArray formatResponse( QString& responseFormat /Out/ ) const;

File: src/server/qgsfcgiserverresponse.h [hunk 5]
```diff
@@ -61,13 +61,13 @@ class QgsFcgiServerResponse: public QgsServerResponse
...
      */
-    virtual void setDefaultHeaders();
+    void setDefaultHeaders();
 
   private:
     QMap<QString, QString> mHeaders;
     QBuffer                mBuffer;
-    bool                   mFinished;
-    bool                   mHeadersSent;
+    bool                   mFinished    = false;
```
Comment (nyalldawson) [inline src/server/qgsfcgiserverresponse.h]: QGIS standard is just a single space, not aligned

File: src/server/qgsserver.cpp (hunk 1 above)
Comment (elpaso) [inline src/server/qgsserver.cpp]: That's correct, now that the plugin architecture is spread all over the server we can remove some of the conditional compilation switches.
I'll have a look to the other points.

Comment (elpaso) [inline src/server/qgsserver.cpp]: @dmarteau that make me think: you should make sure that the server can still be built with WITH_SERVER_PLUGINS disabled.

Comment (elpaso) [inline src/server/qgsserver.cpp]: Removed the conditional in
https://github.com/qgis/QGIS/commit/0f7dfc92855e24921bb4829125a7999291cd66a1

**Comment by elpaso** [comment]
@dmarteau looks good to me. Just one thing you might want to check: if I remember correctly, in my original implementation the reason to keep a reference to the exception was that I wanted to be able to set exceptions from within a plugin. That idea was never implemented, but you might want to consider that now: a Python plugin filter should be able to access to the exception generated by a service and act accordingly (even by clearing the exception and generating its own alternate response).
It would also be very nice if the Python exceptions generated inside a plugin could bubble up into C++, which is not the case now.

**Review by rldhont — APPROVED** [review]

**Comment by dmarteau** [comment]
@elpaso I have removed the reference to the to the exception because there was no   accesor to it: once set it was written the the output and a boolean indicating occucence of an exception was raised. so its sole purpose. Actually a plugin can still set an exception (exception objects are instanciable to python) and the exception is written to the body). From this,  nothing has changed.

If you want that plugins can examine exception I would s suggest to set a new hook 'onExceptionRaised( ... )' that will pass the exception to the plugins when the exception is catched. Mainly because throwing an exception leaves the normal control flow  and the place where you can deal with it it'is when you catch it.

Now about bubbling exceptions from python to C++, yes I had some thoughts about it,  but I need to look further on this (dealing with exception across language can be tricky).

File: src/server/qgsmapserviceexception.h (hunk 3 above)
Comment (dmarteau) [inline src/server/qgsmapserviceexception.h]: This is the idea  at the end, but this is still heavily used in wfs, wcs:  I plan to remove QgsMapServiceException when all services will be moved as new services modules.

**Comment by dmarteau** [comment]
travis config problems ?

**Comment by rldhont** [comment]
@dmarteau build restarted

**Comment by dmarteau** [comment]
@rldhont @nyalldawson Is master broken ? I can't rebase because the project is not compiling:
```
 /Users/david/Projets/qgis/QGIS/build-debug/src/core/qgsexpression_texts.cpp:3:52: error: no member named
    'gFunctionHelpTexts' in 'QgsExpression'
 QHash<QString, QgsExpression::Help> QgsExpression::gFunctionHelpTexts;
```

**Comment by rldhont** [comment]
@dmarteau travis does not complain...

**Comment by nyalldawson** [comment]
>  Is master broken ? I can't rebase because the project is not compiling:

@dmarteau cleaning your build folder should fix this

**Comment by dmarteau** [comment]
@nyalldawson  yes that fixed it, thx.

**Comment by haubourg** [comment]
Hi, this commit seems to break compiling here (ubuntu 16.04). previous commit passes. Following are also broken. 
I have this:
 

> [ 13%] Building CXX object src/core/CMakeFiles/qgis_core.dir/moc_qgsdataprovider.cpp.o
/home/regis/GITHUB_Perso/QGIS3/QGIS/build/src/core/qgsexpression_texts.cpp:3:52: error: ‘QHash<QString, QgsExpression::Help> QgsExpression::gFunctionHelpTexts’ is not a static data member of ‘class QgsExpression’
 QHash<QString, QgsExpression::Help> QgsExpression::gFunctionHelpTexts;
                                                    ^
/home/regis/GITHUB_Perso/QGIS3/QGIS/build/src/core/qgsexpression_texts.cpp: In static member function ‘static void QgsExpression::initFunctionHelp()’:
/home/regis/GITHUB_Perso/QGIS3/QGIS/build/src/core/qgsexpression_texts.cpp:7:8: error: ‘gFunctionHelpTexts’ was not declared in this scope
   if( !gFunctionHelpTexts.isEmpty() )
        ^
/home/regis/GITHUB_Perso/QGIS3/QGIS/build/src/core/qgsexpression_texts.cpp:10:3: error: ‘gFunctionHelpTexts’ was not declared in this scope
   gFunctionHelpTexts.insert( "$area",
   ^
[ 13%] Building CXX object src/core/CMakeFiles/qgis_core.dir/moc_qgsfeature.cpp.o
src/core/CMakeFiles/qgis_core.dir/build.make:9511 : la recette pour la cible « src/core/CMakeFiles/qgis_core.dir/qgsexpression_texts.cpp.o » a échouée
make[2]: *** [src/core/CMakeFiles/qgis_core.dir/qgsexpression_texts.cpp.o] Erreur 1
make[2]: *** Attente des tâches non terminées....
CMakeFiles/Makefile2:1177 : la recette pour la cible « src/core/CMakeFiles/qgis_core.dir/all » a échouée
make[1]: *** [src/core/CMakeFiles/qgis_core.dir/all] Erreur 2
Makefile:160 : la recette pour la cible « all » a échouée
make: *** [all] Erreur 2

**Comment by haubourg** [comment]
please forget, other errors occur later in previous commits too

**Comment by rldhont** [comment]
@haubourg as @nyalldawson has said, you have to clean your build folder.

**Comment by haubourg** [comment]
Ooo, didn't see that thanks!


2017-01-16 14:14 GMT+01:00 rldhont <notifications@github.com>:

**Pull request merged**